sql/grant_tables/grant_tables.go

```diff
@@ -255,8 +290,40 @@ func (g *GrantTables) Negotiate(c *mysql.Conn, user string, addr net.Addr) (mysq
 
 // Persist passes along all changes to the integrator.
 func (g *GrantTables) Persist(ctx *sql.Context) error {
-	//TODO: add the UserPersist interface, using this as a stand-in so I won't forget to put it where it needs to go
-	return nil
+	persistFunc := g.persistFunc
+	if persistFunc == nil {
+		return nil
+	}
+	userEntries := g.user.data.ToSlice(ctx)
+	users := make([]*User, len(userEntries))
+	for i, userEntry := range userEntries {
+		users[i] = userEntry.(*User)
+	}
+	sort.Slice(users, func(i, j int) bool {
+		if users[i].Host == users[j].Host {
+			return users[i].User < users[j].User
+		}
+		return users[i].Host < users[j].Host
+	})
+
+	roleEntries := g.role_edges.data.ToSlice(ctx)
+	roles := make([]*RoleEdge, len(roleEntries))
+	for i, roleEntry := range roleEntries {
+		roles[i] = roleEntry.(*RoleEdge)
+	}
+	sort.Slice(roles, func(i, j int) bool {
```

Comment:
the ordering semantics here could use a comment or a helper function to clarify things

Reply:
I'll leave a comment